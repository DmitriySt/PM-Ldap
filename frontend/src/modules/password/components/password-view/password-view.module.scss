.layout{
  padding: 10px;

}

.editIcon{
  cursor: pointer;
  width: 100%;
  text-align: center;

  &:hover{
    color: #1677ff;
  }
}

.row{
  margin-top: 20px;

}

.label{

  font-weight: bold;

  flex-shrink: 0;

  span{
    cursor: pointer;
  }
}

.value{
  padding-left: 10px;
  overflow-wrap: break-word;
}

.url{
  overflow: hidden;
}

.password{
  padding-left: 0;
}

@media (max-width: 991px) {
  .layout{
    padding-top: 0;
  }
}