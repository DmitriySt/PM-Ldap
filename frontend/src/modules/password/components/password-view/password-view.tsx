import {
  faEllipsisVertical,
  faPenToSquare,
  faTrashCan,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Row, Col, Dropdown, MenuProps, Input, App } from "antd";
import classNames from "classnames";
import { AppError } from "helpers/app-error";
import { IPassword } from "modules/password/types/password-types";
import { FC } from "react";
import styles from "./password-view.module.scss";

export interface PasswordViewProps {
  password: IPassword;
  onEdit: () => void;
  onDelete: () => void;
}

const itemLayout = {
  labelCol: {
    xs: { span: 24 },
    xl: { span: 6 },
  },
  wrapperCol: {
    xs: { span: 24 },
    xl: { span: 18 },
  },
};

export const PasswordView: FC<PasswordViewProps> = ({
  password,
  onEdit,
  onDelete,
}) => {
  const { notification } = App.useApp();

  const menu: MenuProps["items"] = [
    {
      key: "edit",
      label: "Редактировать",
      icon: <FontAwesomeIcon icon={faPenToSquare} />,
      onClick: onEdit,
    },
    {
      key: "delete",
      label: "Удалить",
      icon: <FontAwesomeIcon icon={faTrashCan} />,
      onClick: onDelete,
    },
  ];

  const copyToClipboard = async (value: string, message: string) => {
    try {
      await navigator.clipboard.writeText(value);
      notification.destroy();
      notification.success({ message, duration: 1 });
    } catch (e) {
      notification.error({ message: "Ошибка копирования в буфер." });
      console.log(AppError(e));
    }
  };
  return (
    <div className={styles.layout}>
      <Row wrap={false}>
        <Col flex={"auto"}>
          <h1>{password.name}</h1>
        </Col>

        <Col flex="15px">
          <Dropdown
            menu={{ items: menu }}
            trigger={["hover"]}
            destroyPopupOnHide={false}
          >
            <div className={styles.editIcon}>
              <FontAwesomeIcon icon={faEllipsisVertical} />
            </div>
          </Dropdown>
        </Col>
      </Row>
      <Row className={styles.row}>
        <Col className={styles.label} {...itemLayout.labelCol}>
          <span
            onClick={() =>
              copyToClipboard(
                password.login,
                `Имя пользователя "${password.name}" скопировано в буфер`,
              )
            }
          >
            Имя пользователя:
          </span>
        </Col>
        <Col className={styles.value}>{password.login}</Col>
      </Row>
      <Row className={styles.row}>
        <Col className={styles.label} {...itemLayout.labelCol}>
          <span
            onClick={() =>
              copyToClipboard(
                password.password,
                `Пароль "${password.name}" скопирован в буфер`,
              )
            }
          >
            Пароль:
          </span>
        </Col>
        <Col className={styles.value}>
          <Input.Password
            value={password.password}
            className={styles.password}
            bordered={false}
          />
        </Col>
      </Row>
      <Row className={styles.row}>
        <Col className={styles.label} {...itemLayout.labelCol}>
          <span
            onClick={() =>
              copyToClipboard(
                password.url,
                `URL "${password.name}" скопирован в буфер`,
              )
            }
          >
            URL:
          </span>
        </Col>
        <Col
          className={classNames(styles.value, styles.url)}
          {...itemLayout.wrapperCol}
        >
          <a href={password.url} target="_blank" rel="noreferrer">
            {password.url}
          </a>
        </Col>
      </Row>
      <Row className={styles.row}>
        <Col className={styles.label} {...itemLayout.labelCol}>
          <span
            onClick={() =>
              copyToClipboard(
                password.groupName,
                `Группа "${password.name}" скопирована в буфер`,
              )
            }
          >
            Группа:
          </span>
        </Col>
        <Col className={styles.value}>{password.groupName}</Col>
      </Row>
      <Row className={styles.row}>
        <Col className={styles.label} {...itemLayout.labelCol}>
          <span
            onClick={() =>
              copyToClipboard(
                password.description,
                `Описание "${password.name}" скопировано в буфер`,
              )
            }
          >
            Описание:
          </span>
        </Col>
        <Col className={styles.value}>{password.description}</Col>
      </Row>
    </div>
  );
};
