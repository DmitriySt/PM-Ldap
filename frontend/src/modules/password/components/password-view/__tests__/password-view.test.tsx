import { render, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { PasswordView } from "modules/password/components/password-view/password-view";
import { screen } from "@testing-library/react";
import { passwordStub } from "modules/password/tests/password-group.stub";

afterEach(() => {
  jest.restoreAllMocks();
});

const onEditHandler = jest.fn();
const onDeleteHandler = jest.fn();

describe("password-view", () => {
  it("renders correctly", async () => {
    const view = render(
      <PasswordView
        password={passwordStub}
        onEdit={onEditHandler}
        onDelete={onDeleteHandler}
      />,
    );

    await waitFor(() => {
      expect(screen.getByText(passwordStub.login)).toBeInTheDocument();
    });
    expect(screen.getByDisplayValue(passwordStub.password)).toBeInTheDocument();
    expect(screen.getByText(passwordStub.url)).toBeInTheDocument();
    expect(screen.getByText(passwordStub.description)).toBeInTheDocument();
    expect(view).toMatchSnapshot();

    // eslint-disable-next-line testing-library/no-node-access
    const dropDownMenu = document.querySelector("div.ant-dropdown-trigger");
    expect(dropDownMenu).not.toBeNull();

    await userEvent.hover(dropDownMenu as Element);
    await waitFor(() => {
      expect(screen.getByText("Редактировать")).toBeInTheDocument();
    });
    await userEvent.click(screen.getByText("Редактировать"));
    expect(onEditHandler).toHaveBeenCalledTimes(1);

    await userEvent.hover(dropDownMenu as Element);
    await waitFor(() => {
      expect(dropDownMenu).toHaveClass("ant-dropdown-open");
    });

    await userEvent.click(screen.getByText("Удалить"));
    expect(onDeleteHandler).toHaveBeenCalledTimes(1);
  });
});
