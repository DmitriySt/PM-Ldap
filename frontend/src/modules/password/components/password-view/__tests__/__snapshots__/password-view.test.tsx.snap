// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`password-view renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="layout"
      >
        <div
          class="ant-row ant-row-no-wrap css-dev-only-do-not-override-dkbvqv"
        >
          <div
            class="ant-col css-dev-only-do-not-override-dkbvqv"
            style="flex-basis: auto; min-width: 0;"
          >
            <h1>
              name1
            </h1>
          </div>
          <div
            class="ant-col css-dev-only-do-not-override-dkbvqv"
            style="flex: 0 0 15px; min-width: 0;"
          >
            <div
              class="ant-dropdown-trigger editIcon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-ellipsis-vertical "
                data-icon="ellipsis-vertical"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 128 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="ant-row row css-dev-only-do-not-override-dkbvqv"
        >
          <div
            class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
          >
            <span>
              Имя пользователя:
            </span>
          </div>
          <div
            class="ant-col value css-dev-only-do-not-override-dkbvqv"
          >
            login1
          </div>
        </div>
        <div
          class="ant-row row css-dev-only-do-not-override-dkbvqv"
        >
          <div
            class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
          >
            <span>
              Пароль:
            </span>
          </div>
          <div
            class="ant-col value css-dev-only-do-not-override-dkbvqv"
          >
            <span
              class="ant-input-affix-wrapper ant-input-password password css-dev-only-do-not-override-dkbvqv ant-input-affix-wrapper-borderless css-dev-only-do-not-override-dkbvqv"
            >
              <input
                class="ant-input ant-input-borderless css-dev-only-do-not-override-dkbvqv"
                type="password"
              />
              <span
                class="ant-input-suffix"
              >
                <span
                  aria-label="eye-invisible"
                  class="anticon anticon-eye-invisible ant-input-password-icon"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="eye-invisible"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                    />
                    <path
                      d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </div>
        </div>
        <div
          class="ant-row row css-dev-only-do-not-override-dkbvqv"
        >
          <div
            class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
          >
            <span>
              URL:
            </span>
          </div>
          <div
            class="ant-col value url ant-col-xs-24 ant-col-xl-18 css-dev-only-do-not-override-dkbvqv"
          >
            <a
              href="url1"
              rel="noreferrer"
              target="_blank"
            >
              url1
            </a>
          </div>
        </div>
        <div
          class="ant-row row css-dev-only-do-not-override-dkbvqv"
        >
          <div
            class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
          >
            <span>
              Группа:
            </span>
          </div>
          <div
            class="ant-col value css-dev-only-do-not-override-dkbvqv"
          >
            Мои пароли
          </div>
        </div>
        <div
          class="ant-row row css-dev-only-do-not-override-dkbvqv"
        >
          <div
            class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
          >
            <span>
              Описание:
            </span>
          </div>
          <div
            class="ant-col value css-dev-only-do-not-override-dkbvqv"
          >
            description1
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="layout"
    >
      <div
        class="ant-row ant-row-no-wrap css-dev-only-do-not-override-dkbvqv"
      >
        <div
          class="ant-col css-dev-only-do-not-override-dkbvqv"
          style="flex-basis: auto; min-width: 0;"
        >
          <h1>
            name1
          </h1>
        </div>
        <div
          class="ant-col css-dev-only-do-not-override-dkbvqv"
          style="flex: 0 0 15px; min-width: 0;"
        >
          <div
            class="ant-dropdown-trigger editIcon"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis-vertical "
              data-icon="ellipsis-vertical"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 128 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="ant-row row css-dev-only-do-not-override-dkbvqv"
      >
        <div
          class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
        >
          <span>
            Имя пользователя:
          </span>
        </div>
        <div
          class="ant-col value css-dev-only-do-not-override-dkbvqv"
        >
          login1
        </div>
      </div>
      <div
        class="ant-row row css-dev-only-do-not-override-dkbvqv"
      >
        <div
          class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
        >
          <span>
            Пароль:
          </span>
        </div>
        <div
          class="ant-col value css-dev-only-do-not-override-dkbvqv"
        >
          <span
            class="ant-input-affix-wrapper ant-input-password password css-dev-only-do-not-override-dkbvqv ant-input-affix-wrapper-borderless css-dev-only-do-not-override-dkbvqv"
          >
            <input
              class="ant-input ant-input-borderless css-dev-only-do-not-override-dkbvqv"
              type="password"
            />
            <span
              class="ant-input-suffix"
            >
              <span
                aria-label="eye-invisible"
                class="anticon anticon-eye-invisible ant-input-password-icon"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="eye-invisible"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M942.2 486.2Q889.47 375.11 816.7 305l-50.88 50.88C807.31 395.53 843.45 447.4 874.7 512 791.5 684.2 673.4 766 512 766q-72.67 0-133.87-22.38L323 798.75Q408 838 512 838q288.3 0 430.2-300.3a60.29 60.29 0 000-51.5zm-63.57-320.64L836 122.88a8 8 0 00-11.32 0L715.31 232.2Q624.86 186 512 186q-288.3 0-430.2 300.3a60.3 60.3 0 000 51.5q56.69 119.4 136.5 191.41L112.48 835a8 8 0 000 11.31L155.17 889a8 8 0 0011.31 0l712.15-712.12a8 8 0 000-11.32zM149.3 512C232.6 339.8 350.7 258 512 258c54.54 0 104.13 9.36 149.12 28.39l-70.3 70.3a176 176 0 00-238.13 238.13l-83.42 83.42C223.1 637.49 183.3 582.28 149.3 512zm246.7 0a112.11 112.11 0 01146.2-106.69L401.31 546.2A112 112 0 01396 512z"
                  />
                  <path
                    d="M508 624c-3.46 0-6.87-.16-10.25-.47l-52.82 52.82a176.09 176.09 0 00227.42-227.42l-52.82 52.82c.31 3.38.47 6.79.47 10.25a111.94 111.94 0 01-112 112z"
                  />
                </svg>
              </span>
            </span>
          </span>
        </div>
      </div>
      <div
        class="ant-row row css-dev-only-do-not-override-dkbvqv"
      >
        <div
          class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
        >
          <span>
            URL:
          </span>
        </div>
        <div
          class="ant-col value url ant-col-xs-24 ant-col-xl-18 css-dev-only-do-not-override-dkbvqv"
        >
          <a
            href="url1"
            rel="noreferrer"
            target="_blank"
          >
            url1
          </a>
        </div>
      </div>
      <div
        class="ant-row row css-dev-only-do-not-override-dkbvqv"
      >
        <div
          class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
        >
          <span>
            Группа:
          </span>
        </div>
        <div
          class="ant-col value css-dev-only-do-not-override-dkbvqv"
        >
          Мои пароли
        </div>
      </div>
      <div
        class="ant-row row css-dev-only-do-not-override-dkbvqv"
      >
        <div
          class="ant-col label ant-col-xs-24 ant-col-xl-6 css-dev-only-do-not-override-dkbvqv"
        >
          <span>
            Описание:
          </span>
        </div>
        <div
          class="ant-col value css-dev-only-do-not-override-dkbvqv"
        >
          description1
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
